feat(#65): get lints for gh-pages in runtime

Marat-Tim opened the pull request:
adds all motives of active lints to the Github pages branch

Marat-Tim commented:
I understand that there are many errors here, but I wanted to do something so that you could specifically point them out to me

yegor256 commented:
@Marat-Tim looks like a good start! Now, please, try to make all build jobs pass.

Marat-Tim commented:
@yegor256 I'm ready for my code to be destroyed)

I have a few questions

1. Do we collect the jar file somewhere and can we download it from there? Then from pom.xml it will be possible to remove the build plugin in jar

2. I assume that the PrintAllLines class needs to be moved to a separate project. If I create a project in the repository, won't it break some kind of pipeline? And again, how can I connect the latest version of the main project to subproject?

3. On behalf of whom to make a push request to the gh-pages branch? Now I've done it on my own login

4. How good is the idea of JavaLint? After all, if I wrap it at the wrong moment (for example, after wrapping it in Unlint), the name will be lost. Or you'll have to recursively get the decorated object, which sounds pretty crooked

Marat-Tim commented:
@yegor256 I updated and corrected the pull request. An example of how the gh-pages branch will be created with it can be viewed [here](https://github.com/Marat-Tim/lints/tree/gh-pages)

yegor256 commented:
@Marat-Tim the `SaveAll.java` file is not in the right place. It's among feature files, which are shipped to the client in the final JAR. Should it be there? Maybe we can replace it with a Bash script?

yegor256 commented:
@h1alexbel please, take a lookt

Marat-Tim commented:
@yegor256 We can't replace java code with bash if we want to get a list of motives in java runtime. I also tried to move this executable file to a separate subproject (similar to lints-it), but I ran into the problem that `PkWpa` and `PkMono` are not public classes, so the only way to use them is to write a file in the main project

Would it be better to just exclude this file from the build into the final jar? I just don't really understand which plugin this assembly is configured in, if you tell me, it will be very good for me.

h1alexbel commented on pom.xml:
```diff
@@ -243,7 +243,9 @@
             <configuration>
               <excludes>
                 <exclude>pmd:/src/it/.*</exclude>
+                <exclude>pmd:/src/main/java/org/eolang/lints/SaveAll.java</exclude>
```
@Marat-Tim Why exclude this class from quality analysis?

Marat-Tim replied:
@h1alexbel I thought, since this is a script, then we don't need to check it for quality

By analogy with classes in the `lints-it` subproject

h1alexbel commented on src/main/java/org/eolang/lints/SaveAll.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.stream.Stream;
+import org.cactoos.iterable.Joined;
+
+public class SaveAll {
```
@Marat-Tim This Java main class won't fit us. Ideally, we need some script that won't be shipped to the client, but can be executed during the build and forgotten. I thought about the following setup: 
1. Create script in Groovy language. with Java semantics support (we will get all the `Lint` interface implementations and call `name()` method there). For XSL lints names, we don't need Java, since we can iterate over `src/main/resources/org/eolang/lints/**.xsl` to collect them.
2. Run the script during the build with the help of [GMaven](https://groovy.github.io/GMavenPlus) plugin
3. Place the results to the `README.md` under `## Lints` section.

@yegor256 WDYT about Groovy script?

Marat-Tim replied:
@h1alexbel we cannot get a list of all names by searching through folders with xsl files, since not all lints are implemented on xsl. In theory, we can go through the folder with markdown files, but then we need to write a test that checks that we don't have unnecessary markdown files without implementing the appropriate lint

What's the problem with keeping my class and excluding it from the final jar build?

Marat-Tim replied:
@h1alexbel @yegor256 Now, as far as I understand, the `motive` method is not used at all (at least its behavior is very strange, somewhere it returns the path to the file in jar, somewhere the text of the file from jar)

Is there no such thing that it should return a full-fledged markdown document (not the path to it)? In this approach, we will be able to dynamically create documentation using data from runtime

h1alexbel replied:
@Marat-Tim I meant we can aggregate the list of lints this way:
* find XSL lints from FS
* find all Java lints

All of these manipulations can be listed in the Groovy script. Thus, you won't need to place this Java Utility class with our production classes, and we will keep [SoC](https://en.wikipedia.org/wiki/Separation_of_concerns) in the project.

Marat-Tim replied:
@h1alexbel We can transfer this class to another package if we need to separate it from the production code

I just don't quite get it, how do we get a list of Java lints in groovy?

yegor256 replied:
@Marat-Tim @h1alexbel indeed, Groovy is a good option for such a script

h1alexbel commented on .github/motives.groovy:
```diff
@@ -0,0 +1,18 @@
+ * SPDX-FileCopyrightText: Copyright (c) 2016-2025 Objectionary.com
+ * SPDX-License-Identifier: MIT
+ */
+
```
@Marat-Tim let's remove this empty line

h1alexbel commented on pom.xml:
```diff
@@ -534,6 +534,23 @@
+        <version>4.1.1</version>
+        <configuration>
+          <scripts>
+            <script>./.github/motives.groovy</script>
```
@Marat-Tim let's move the script into `src/main/groovy/org/eolang/lints/motives.groovy`

h1alexbel replied:
@Marat-Tim in the `jcabi-parent`, we have this `groovy` profile:

```xml
<profile>
    <!--
    Compile Groovy sources from src/main/groovy and src/test/groovy
    directories.
    -->
    <id>groovy</id>
    <activation>
      <file>
        <exists>${basedir}/src/main/groovy</exists>
      </file>
    </activation>
    <dependencies>
      <dependency>
        <groupId>org.codehaus.groovy</groupId>
        <artifactId>groovy-all</artifactId>
        <version>3.0.24</version>
        <type>pom</type>
      </dependency>
    </dependencies>
    <build>
      <plugins>
        <plugin>
          <groupId>org.codehaus.gmavenplus</groupId>
          <artifactId>gmavenplus-plugin</artifactId>
          <version>4.2.0</version>
          <configuration>
            <verbose>true</verbose>
          </configuration>
          <executions>
            <execution>
              <id>jcabi-groovy</id>
              <goals>
                <goal>compile</goal>
                <goal>compileTests</goal>
              </goals>
            </execution>
          </executions>
        </plugin>
      </plugins>
    </build>
  </profile>
```

h1alexbel replied:
@Marat-Tim correct. Your Groovy scripts should be located in the `target/classes`. From there you can invoke them using `gmavenplus-plugin` passing path to the script:

```xml
<scripts>
  <script>src/main/groovy/org/eolang/lints/SaveMotives.groovy</script>
</scripts>
```

h1alexbel commented on .github/motives.groovy:
```diff
@@ -0,0 +1,18 @@
+
+package org.eolang.lints
+
+static def saveLint(Lint<?> lint) {
```
@Marat-Tim let's add JavaDoc here, explaining why this script exists

h1alexbel commented on .github/motives.groovy:
```diff
@@ -0,0 +1,18 @@
+
+package org.eolang.lints
+
+static def saveLint(Lint<?> lint) {
```
@Marat-Tim maybe we can setup tests for `saveLint()` method, in the `src/test/groovy`? Otherwise, we don't and _won't_ know if we break something during the future modifications. WDYT?

Marat-Tim replied:
@h1alexbel when I tried to do this, I got an error https://github.com/jcabi/jcabi-parent/issues/586

I added a puzzle to this fix

h1alexbel reviewed (changes requested):
@Marat-Tim check my comments, please

Marat-Tim committed:
feat(#65): add docs to README + get motives at runtime

Marat-Tim commented:
@h1alexbel

maxonfjvipon commented:
@h1alexbel please check this one

h1alexbel commented on .github/motives.groovy:
```diff
@@ -0,0 +1,27 @@
+ * This script saves a file with the lint name and extension .md to the tmp folder.
+ * The motive of this lint is preserved as content.
+ * These files are saved for all active lints in the project.
+ * @todo #65:30min after fixing this problem with creation of groovy folder
```
@Marat-Tim let's make two todos for these issues:
* `"Move this script to the src/main/groovy folder."`
* `"Write a test for motives.groovy"`

h1alexbel reviewed (changes requested):
@Marat-Tim just one comment, please check

h1alexbel reviewed (approved):
@Marat-Tim LGTM

h1alexbel commented:
@yegor256 take a look, please

Marat-Tim commented:
@yegor256

yegor256 reviewed (changes requested):
a new version of jcabi-parent is available

Marat-Tim commented:
@yegor256 Yes, I'm working on it right now - but I've got a question: https://github.com/objectionary/lints/pull/137#discussion_r2085181413
